Spanner DBAL driver.

julien-sebire opened the pull request:
**Goal:**
implementation of spanner dbal driver 

**Status:**
please check the status inside this :
https://oat-sa.atlassian.net/wiki/spaces/NEX/pages/191828848/Integrate+Spanner+implementation+to+package-tao

Required by https://github.com/oat-sa/generis/pull/629

julien-sebire committed:
Added implementation for now expression to store to TIMESTAMP datatype.

ekkinox commented on src/SpannerPlatform.php:
```diff
@@ -72,6 +72,13 @@ public function getBlobTypeDeclarationSQL(array $field)
+    {
+        $dateTime = new \DateTime('now', new \DateTimeZone('UTC'));
+
+        return ($dateTime)->format('U');
```
`return $dateTime->format('U');`

julien-sebire committed:
Removed useless parenthesis.

julien-sebire committed:
Deactivated schema modification queries execution on spanner.

julien-sebire committed:
Fixed dateTime format.
Disabled DDL statements run by "query" instead of "exec".

julien-sebire committed:
Set driver name as a constant for external comparisons.

julien-sebire committed:
Separated spanner connection and connection wrapper. Has to be cleaned.

julien-sebire committed:
Provided custom platform to schemaManager.

julien-sebire committed:
Added credential file for spanner client initialization.

julien-sebire committed:
More explicit exception message on spanner query error.

julien-sebire committed:
Explicitely give instance name to list database.

julien-sebire committed:
Custom platform additions (now expression, Doctrine types, schema reading).

julien-sebire committed:
Added optional Spanner instance name as parameter for ListDatabases.

julien-sebire committed:
Used Spanner native functions for DML instead of generating native queries.

julien-sebire committed:
Added cache for SQL statement parsing.

julien-sebire committed:
Fixed SQL statement parsing for parameter names replacement.

julien-sebire committed:
Fixed rows reading with generator.
TODO: This will have to be improved since we now read all rows at once.

julien-sebire committed:
Seperated execution of DML and SQL statements + fixed passing query parameters.

julien-sebire committed:
Required phpmyadmin/sql-parser for Sql statement parsing.

julien-sebire committed:
Fixed platform's datetime format + removed useless getNowExpression.

julien-sebire committed:
Removed useless parameters to SpannerConnection constructor.

julien-sebire committed:
Used a constant for spanner driver and platform names.

julien-sebire committed:
Fixed tests for parameter place holders conversion.

julien-sebire committed:
Added support for FETCH_OBJ hydration.

julien-sebire committed:
Removed duplicated reading of result set from spanner client generator.

julien-sebire committed:
Added support for positional parameters shifted index offset (coming from DBAL statement parametrized query translated to positional with parameter types).

julien-sebire committed:
Allowed passing DML query result through rowCount() instead of direct return (needed by DBAL\Connection::executeUpdate).

julien-sebire committed:
Cleaned up Connection and ConnectionWrapper.

julien-sebire committed:
Separated parameters translation in Statement.

julien-sebire committed:
First unit tests for Statement.

siwane committed:
Add unit test for SpannerKeywords class

julien-sebire committed:
Added copyright notes.

siwane committed:
Merge branch 'feature/NEX-84/spanner-query-refactoring' of https://github.com/oat-sa/lib-dbal-spanner into feature/NEX-84/spanner-query-refactoring

julien-sebire committed:
Injected driver in Connection creation to retrieve Platform when needed.

siwane committed:
Merge branch 'feature/NEX-84/spanner-query-refactoring' of https://github.com/oat-sa/lib-dbal-spanner into feature/NEX-84/spanner-query-refactoring

siwane committed:
Update README

siwane committed:
Check if identifiers are not empty when forge where clause

siwane committed:
Add unit test for SpannerConnection class

siwane committed:
Merge branch 'feature/NEX-84/spanner-query-refactoring' of https://github.com/oat-sa/lib-dbal-spanner into feature/NEX-84/spanner-query-refactoring

siwane committed:
Add test for exec with ddl query

siwane committed:
Add test coverage for NullExpression

siwane committed:
Fix typo

julien-sebire committed:
Added unit tests for SpannerStatement and ParameterTranslator.

julien-sebire committed:
Merge branch 'feature/NEX-84/spanner-query-refactoring' of github.com:oat-sa/lib-dbal-spanner into feature/NEX-84/spanner-query-refactoring

julien-sebire committed:
Simplified SpannerConnectionWrapper to a simple proxy to SpannerConnection. Not to be confused with SpannerConnection, which embeds a database and a lot more code. These actually are two separated objects.

siwane committed:
Add unit test for SpannerDriver

siwane committed:
Merge branch 'feature/NEX-84/spanner-query-refactoring' of https://github.com/oat-sa/lib-dbal-spanner into feature/NEX-84/spanner-query-refactoring

siwane committed:
Add unit test for SpannerConnectionWrapper

julien-sebire committed:
Completed SpannerStatement testing.

julien-sebire committed:
Merge branch 'feature/NEX-84/spanner-query-refactoring' of github.com:oat-sa/lib-dbal-spanner into feature/NEX-84/spanner-query-refactoring

julien-sebire committed:
Added tests for Platform and SchemaManager.

julien-sebire committed:
Full unit tests.

julien-sebire committed:
Moved spanner client factory test to integration testing.

julien-sebire committed:
Removed integration test suite from phpunit.xml.

julien-sebire committed:
Added authentication cache.
Created cache session pool.
TODO: Cache session pool has to be used in database connection for multiple sessions to share authentication cache in SpannerDriver::connect.

julien-sebire committed:
NEX-371: fixed DDL execution.

julien-sebire committed:
Added optional injection of spannerClientFactory. Driver creation is called by Dbal DriverManager, so injection is not possible.

julien-sebire committed:
Default is not supported on Spanner.

julien-sebire committed:
TEXT data type is translated to blob, which doesn't match string type, so back to string(max) (max size).

boajer committed:
Safer logging ( no more Error if no logger injected into object)
FQN simplified

julien-sebire committed:
Read credential filename from .env file + fixed session pool

mike-ionut-mihai-sandu-tao commented on composer.json:
```diff
@@ -13,10 +13,12 @@
         "google/cloud-spanner": "^1.15",
-        "grpc/grpc": "^1.19"
+        "grpc/grpc": "^1.19",
+        "phpmyadmin/sql-parser": "dev-master"
```
why do we need this?

julien-sebire replied:
To parse the parameters in SQL statements in the form of :param1.
I'd be glad to use another library to do it if we can find a faster one.
I won't go to writing regular expressions for it.

boajer replied:
Ca we use  fixed version to something from https://github.com/phpmyadmin/sql-parser/releases, but not directly to master?

mike-ionut-mihai-sandu-tao replied:
done : 
"phpmyadmin/sql-parser": "^5.2.0"

mike-ionut-mihai-sandu-tao commented on phpunit.xml:
```diff
@@ -3,8 +3,11 @@
+
+    <filter>
+        <whitelist>
+            <directory>src</directory>
```
why are we forcefully excluding the `src`  folder

julien-sebire replied:
We're indeed **including** it.

mike-ionut-mihai-sandu-tao commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,120 @@
+class ParameterTranslator
+{
+    /** @var int */
+    protected $positionalParameterCount = 0;
```
if we don't plan to extend this class, let change the visibility to private, please

julien-sebire replied:
Done

mike-ionut-mihai-sandu-tao commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,120 @@
+     */
+    public function translatePlaceHolders(string $sql): string
+    {
+        if (strpos($sql, ':') === false && strpos($sql, '?') === false) {
```
[Optional] Yoda ?

julien-sebire replied:
Why?
https://dev.to/greg0ire/why-using-yoda-conditions-you-should-probably-not

mike-ionut-mihai-sandu-tao commented on .phpunit.result.cache:
```diff
@@ -0,0 +1 @@
+C:37:"PHPUnit\Runner\DefaultTestResultCache":17722:{a:2:{s:7:"defects";a:0:{}s:5:"times";a:168:{s:105:"Oat\DbalSpanner\Tests\Unit\Parameters\ParameterTranslatorTest::testTranslatePlaceHolders with data set #0";d:0.006;s:105:"Oat\DbalSpanner\Tests\Unit\Parameters\ParameterTranslatorTest::testTranslatePlaceHolders with data set #1";d:0.032;s:105:"Oat\DbalSpanner\Tests\Unit\Parameters\ParameterTrans ...
```
should this be here? i think we should not commit this file

julien-sebire committed:
Merge pull request #3 from oat-sa/fix/NEX-422/error_logging

Failsafe logging

julien-sebire committed:
Added message whe google credentials filename is missing.

julien-sebire committed:
Merge branch 'feature/NEX-84/spanner-query-refactoring' of github.com:oat-sa/lib-dbal-spanner into feature/NEX-84/spanner-query-refactoring

boajer committed:
Introduced types translation

boajer committed:
Fixed unittests

boajer committed:
Fixed exception throwing

boajer committed:
change reverted

julien-sebire committed:
Merge pull request #4 from oat-sa/fix/NEX-422/dbal_to_spanner_type_mapping

Fix/nex 422/dbal to spanner type mapping

julien-sebire committed:
Added default credentials filename.

julien-sebire committed:
Merge branch 'feature/NEX-84/spanner-query-refactoring' of github.com:oat-sa/lib-dbal-spanner into feature/NEX-84/spanner-query-refactoring

boajer committed:
Temporary empty implementation of some unsupported features

mike-ionut-mihai-sandu-tao commented on composer.json:
```diff
@@ -13,10 +13,12 @@
         "google/protobuf": "^3.7",
         "google/cloud-spanner": "^1.15",
-        "grpc/grpc": "^1.19"
+        "grpc/grpc": "^1.19",
```
we just found out that this is not necessary 
`"grpc/grpc": "^1.19",`
`"google/protobuf": "^3.7",`
we just need the installed version
also when you install `"google/cloud-spanner": "^1.15",` it will download the grpc and protobuf lib

mike-ionut-mihai-sandu-tao commented on src/SpannerDriver.php:
```diff
@@ -67,22 +105,22 @@ public function getDatabase(Connection $conn)
+    public function selectDatabase(string $databaseName, array $options): Database
     {
-        if (!in_array($databaseName, $this->showDatabases())) {
+        if (!in_array($databaseName, $this->listDatabases($this->instance->name()))) {
```
i have just removed this while testing and it's making the response time way smaller 
why do we need to list the databases on the each connection that has not been warmed up with the cache pool

mike-ionut-mihai-sandu-tao commented on src/SpannerKeywords.php:
```diff
@@ -0,0 +1,142 @@
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
```
this should be at the beginning of the file and not after the license

julien-sebire replied:
Don't think so. See the example in here: https://www.php-fig.org/psr/psr-12/

julien-sebire committed:
Merge pull request #5 from oat-sa/fix/NEX-519/transactions_and_lock_workaround

Temporary empty implementation of some unsupported features

mike-ionut-mihai-sandu-tao committed:
removed protobuf and grpc php libs,  this should be based on the extension.so

mike-ionut-mihai-sandu-tao committed:
removed the list databases function for connect

julien-sebire committed:
Fixed tests.

boajer committed:
Improved PDO compatibility

julien-sebire committed:
Merge pull request #6 from oat-sa/bugfix/NEX-422/issues-related-to-session

Bugfix/nex 422/issues related to session + fixing tests not updated in previous commits.

julien-sebire committed:
Backported listDatabase function used in tao installation.

julien-sebire committed:
Added transactional support.

julien-sebire committed:
Fixed tests.

julien-sebire committed:
Removed useless connectionWrapper.

julien-sebire committed:
Added types.

julien-sebire committed:
Code style.

julien-sebire committed:
Moved test database to existing camera-obscura project.

julien-sebire committed:
Added support for transactions.

mike-ionut-mihai-sandu-tao committed:
removed the extra check that was generating huge load on spanner with the getinstance call

siwane committed:
Merge pull request #7 from oat-sa/bugfix/remove-the-getInstance-bottleneck

removed the extra check that was generating huge load on spanner with…

julien-sebire committed:
Fixes for PR.

julien-sebire reviewed (commented):
Answered comments.

siwane committed:
Fix unit test & escape query

siwane committed:
Adapt unit test for new behaviour (no instance->exists() anymore)

mike-ionut-mihai-sandu-tao reviewed (approved):
i think this is fit for a pre-alfa release,
we have pointed some improvements and they will be added the the spike document that contains the status of spanner

mike-ionut-mihai-sandu-tao commented:
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

boajer commented on composer.json:
```diff
@@ -11,12 +11,12 @@
     },
     "require-dev":  {
-        "phpunit/phpunit": "^8.1"
+        "phpunit/phpunit": "^8.1",
```
Wouldn't it conflict with a https://github.com/oat-sa/generis/pull/735/files#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780R75 ?

mike-ionut-mihai-sandu-tao replied:
we can leave it there as this does not affect the root composer usage

bartlomiejmarszal commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,140 @@
+     * DBAL and Spanner are both named so can be compatible.
+     * Named and positional syntaxes are not compatible.
+     *
+     * @param string $sql
```
I don't think that we need this in PHP docs if we have type hinting for this param and you are not including any additional info for this param. 
```suggestion
```

mike-ionut-mihai-sandu-tao replied:
this will be as a clean-up , 
please gather all of the improvements ideas and put them inside the document listed inside the description of the PR

boajer reviewed (approved):
This is good enough for alpha version, it has been used on various machines already for a while, but for sure we should collect points for improving this later

bartlomiejmarszal commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,140 @@
+
+    /**
+     * Translates the positional parameters into named parameters.
+     * Takes care about types mapp
```
```suggestion
     * Takes care about types map
```

bartlomiejmarszal commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,140 @@
+     * Translates the positional parameters into named parameters.
+     * Takes care about types mapp
+     *
+     * @param array $boundValues
```
Parameters already type hinted in method signature
```suggestion
```

bartlomiejmarszal commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,140 @@
+     * Takes care about types mapp
+     *
+     * @param array $boundValues
+     * @param array|null $params
```
Parameters already type hinted in method signature
```suggestion
```

bartlomiejmarszal commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,140 @@
+     * @param array $boundValues
+     * @param array|null $params
+     *
+     * @param array|null $boundTypes
```
Parameters already type hinted in method signature
```suggestion
```

bartlomiejmarszal commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,140 @@
+     * @param array|null $params
+     *
+     * @param array|null $boundTypes
+     * @return array
```
return already type hinted in the method signature
```suggestion
```

bartlomiejmarszal commented on tests/Unit/Parameters/ParameterTranslatorTest.php:
```diff
@@ -0,0 +1,132 @@
+     */
+    public function testConvertPositionalToNamedWithNamedParameters(array $expected, array $boundValues, array $params = null)
+    {
+        $this->assertEquals($expected, $this->subject->convertPositionalToNamed($boundValues, $params));
```
I find this way of representing unit tests more readable. 
```suggestion
        $result = $this->subject->convertPositionalToNamed($boundValues, $params);
        $this->assertEquals($expected, $result);
```

mike-ionut-mihai-sandu-tao committed:
modified the phpmyadmin/sql-parser composer version

mike-ionut-mihai-sandu-tao committed:
phpcs applied

bartlomiejmarszal commented on tests/Unit/Parameters/ParameterTranslatorTest.php:
```diff
@@ -0,0 +1,133 @@
+    /**
+     * @dataProvider positionalParamsToTest
+     */
+    public function testConvertPositionalToNamedWithPositionalParameters(array $expected, array $boundValues, array $params = null)
```
Why don't you test `boundTypes` param?

bartlomiejmarszal commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,140 @@
+     * @param $type
+     * @return int|mixed
+     */
+    public function convertPDOtoSpannerTypes($type)
```
Missing type hint for param.

bartlomiejmarszal replied:
This method should be unit tested as well

bartlomiejmarszal commented on src/Parameters/ParameterTranslator.php:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    /**
+     * @param $type
```
Missing description

bartlomiejmarszal commented on src/SpannerClient/SpannerClientFactory.php:
```diff
@@ -0,0 +1,83 @@
+        $authCache = new SysVCacheItemPool();
+        $keyFileName = $_ENV[self::KEY_FILE_ENV_VARIABLE] ?? self::DEFAULT_CREDENTIALS_FILE;
+        if ($keyFileName === '') {
+            new GoogleException(
```
Don't you want to throw exception here?

bartlomiejmarszal commented on src/SpannerClient/SpannerClientFactory.php:
```diff
@@ -0,0 +1,83 @@
+    public function create()
+    {
+        $authCache = new SysVCacheItemPool();
+        $keyFileName = $_ENV[self::KEY_FILE_ENV_VARIABLE] ?? self::DEFAULT_CREDENTIALS_FILE;
```
Using `_ENV` is deprecated?

mike-ionut-mihai-sandu-tao replied:
nope

bartlomiejmarszal commented on src/SpannerClient/SpannerClientFactory.php:
```diff
@@ -0,0 +1,83 @@
+        if ($keyFileName === '') {
+            new GoogleException(
+                sprintf(
+                    'Missing path to Google credentials key file (should be set as an environment variable "%s").',
```
It's not covered by unit tests.

mike-ionut-mihai-sandu-tao replied:
please add this as an improvement

bartlomiejmarszal commented on src/SpannerClient/SpannerClientFactory.php:
```diff
@@ -0,0 +1,83 @@
+     * Creates a session pool to allow multiple sessions to share the auth cache.
+     *
+     * @return CacheSessionPool
+     * @throws \Exception
```
Import Exception in use block.
```suggestion
     * @throws Exception
```

bartlomiejmarszal commented on src/SpannerClient/SpannerClientFactory.php:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * Creates a session pool to allow multiple sessions to share the auth cache.
+     *
+     * @return CacheSessionPool
```
Why don't we use return type?

bartlomiejmarszal commented on src/SpannerClient/SpannerClientFactory.php:
```diff
@@ -0,0 +1,83 @@
+     * @return CacheSessionPool
+     * @throws \Exception
+     */
+    public function createCacheSessionPool()
```
Missing unit test

bartlomiejmarszal commented on src/SpannerConnection.php:
```diff
@@ -1,41 +1,101 @@
     }
 
+    /**
+     * @param string $prepareString
```
Why don't we use type hint?

bartlomiejmarszal commented on src/SpannerConnection.php:
```diff
@@ -1,41 +1,101 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```

bartlomiejmarszal commented on src/SpannerConnection.php:
```diff
@@ -1,41 +1,101 @@
+            return $this->cachedStatements[$prepareString];
+        }
+
+        $statement = new SpannerStatement($this->database, $prepareString);
```
Missing `\Doctrine\DBAL\Exception\InvalidArgumentException` throw in phpdocs

bartlomiejmarszal commented on src/SpannerConnection.php:
```diff
@@ -1,41 +1,101 @@
 
+    /**
+     * @return SpannerStatement
+     */
```
Missin throw exception tag

bartlomiejmarszal commented on src/SpannerConnection.php:
```diff
@@ -1,41 +1,101 @@
         $args = func_get_args();
         $sql = $args[0];
+        // Runs schema manipulation on the DDL endpoint.
+        if ($this->isDdlStatement($sql)) {
```
Can we modify `isDdlStatement` so it will contain phpdocs or type hinting?

bartlomiejmarszal commented on src/SpannerConnection.php:
```diff
@@ -45,79 +105,111 @@ public function delete($tableExpression, array $identifier, array $types = [])
+    protected function getParameterType($value, $type): int
+    {
+        if ($type !== '') {
+            return $type;
```
This may not return int as it's type hinted.

bartlomiejmarszal commented on src/SpannerConnection.php:
```diff
@@ -127,35 +219,62 @@ public function quote($input, $type = ParameterType::STRING)
+    /**
+     * @param string $statement
+     *
+     * @return int|mixed
```
Method may return bool

bartlomiejmarszal commented on src/SpannerConnectionWrapper.php:
```diff
@@ -0,0 +1,137 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```

bartlomiejmarszal commented on src/SpannerDriver.php:
```diff
@@ -1,20 +1,42 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```

bartlomiejmarszal commented on src/SpannerDriver.php:
```diff
@@ -24,23 +46,46 @@ class SpannerDriver implements Driver
+    /**
+     * SpannerDriver constructor.
+     *
+     * @param SpannerClientFactory $spannerClientFactory
```
```suggestion
```

bartlomiejmarszal commented on src/SpannerDriver.php:
```diff
@@ -50,39 +95,41 @@ public function getDatabasePlatform()
+        return new SpannerSchemaManager($conn, new SpannerPlatform());
     }
 
     public function getName()
```
missing phpdocs or return type

bartlomiejmarszal commented on src/SpannerDriver.php:
```diff
@@ -50,39 +95,41 @@ public function getDatabasePlatform()
 
     public function getSchemaManager(Connection $conn)
```
missing phpdocs or return type

bartlomiejmarszal commented on src/SpannerDriver.php:
```diff
@@ -50,39 +95,41 @@ public function getDatabasePlatform()
+    public function transactional(Closure $closure)
+    {
+        if (!$this->connection instanceof SpannerConnection) {
+            throw new DBALException('Can not run transaction without connecting first.');
```
What if connection would be different instance then SpannerCnnection?

bartlomiejmarszal commented on src/SpannerDriver.php:
```diff
@@ -134,12 +174,19 @@ public function getInstance(string $instanceName): Instance
+     * @throws GoogleException
      */
-    public function showDatabases(): array
+    public function listDatabases(string $instanceName = ''): array
```
The default value for param that should be instances of some class should be `null` 
```suggestion
    public function listDatabases(string $instanceName = null): array
```

bartlomiejmarszal commented on src/SpannerDriver.php:
```diff
@@ -134,12 +174,19 @@ public function getInstance(string $instanceName): Instance
      *
+     * @param string $instanceName
+     *
      * @return array|string[]
```
```suggestion
     * @return string[]
```

bartlomiejmarszal commented on src/SpannerDriver.php:
```diff
@@ -134,12 +174,19 @@ public function getInstance(string $instanceName): Instance
     /**
      * Returns a list of database names existing on a Spanner instance.
      *
+     * @param string $instanceName
```
```suggestion
```

bartlomiejmarszal commented on src/SpannerKeywords.php:
```diff
@@ -0,0 +1,143 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```

bartlomiejmarszal commented on src/SpannerKeywords.php:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
```
return type missing

bartlomiejmarszal commented on src/SpannerSchemaManager.php:
```diff
@@ -1,25 +1,78 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```

bartlomiejmarszal commented on src/SpannerSchemaManager.php:
```diff
@@ -1,25 +1,78 @@
 
 class SpannerSchemaManager extends AbstractSchemaManager
 {
     public function listDatabases()
```
return type missing

bartlomiejmarszal commented on src/SpannerSchemaManager.php:
```diff
@@ -1,25 +1,78 @@
         throw new \Exception("\e[31m\e[1m" . __METHOD__ . "\e[21m\e[0m" . ' not implemented.');
     }
 
     protected function _getPortableTableColumnDefinition($tableColumn)
```
param hint missing

bartlomiejmarszal commented on src/SpannerSchemaManager.php:
```diff
@@ -1,25 +1,78 @@
+    protected function _getPortableTableIndexesList($tableIndexes, $tableName = null)
+    {
+        $convertedIndexes = [];
+        foreach ($tableIndexes as $k => $v) {
```
```suggestion
        foreach ($tableIndexes as $key => $value) {
```

bartlomiejmarszal commented on src/SpannerSchemaManager.php:
```diff
@@ -1,25 +1,78 @@
+        return $table['table_name'];
+    }
+
+    protected function _getPortableTableIndexesList($tableIndexes, $tableName = null)
```
missing param hint or phpdocs

bartlomiejmarszal commented on src/SpannerStatement.php:
```diff
@@ -1,5 +1,23 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```

bartlomiejmarszal commented on src/SpannerStatement.php:
```diff
@@ -173,46 +126,107 @@ public function bindParam($column, &$variable, $type = ParameterType::STRING, $l
+            [$parameters, $types] = $this->parameterTranslator->convertPositionalToNamed($this->boundValues, $params, $this->boundTypes);
+        } catch (InvalidArgumentException $exception) {
+            if ($this->logger) {
+                $this->logger->error(
```
Can we extract this to private method?

bartlomiejmarszal commented on src/SpannerStatement.php:
```diff
@@ -225,38 +239,157 @@ public function setFetchMode($fetchMode, $arg2 = null, $arg3 = null)
+        $lastRow = count($this->rows) - 1;
+
+        switch ($cursorOrientation) {
+            case PDO::FETCH_ORI_NEXT:
```
Can we describe each of those cases?

bartlomiejmarszal commented on src/SpannerStatement.php:
```diff
@@ -225,38 +239,157 @@ public function setFetchMode($fetchMode, $arg2 = null, $arg3 = null)
+     *
+     * @return string|int
+     */
+    public function getRealMode($fetchMode)
```
Why is this public method?

bartlomiejmarszal reviewed (changes requested):
I'm confused about using param or return type hinting. Sometimes it's used sometimes is not. Sometimes you use phpdocs sometimes not. IMO when you use type hinting it's not required to describe this same information in phpdocs. Unless you want to specify something more explicitly.

mike-ionut-mihai-sandu-tao committed:
Update src/Parameters/ParameterTranslator.php

Co-Authored-By: Bartłomiej Marszał <bartlomiej@taotesting.com>

mike-ionut-mihai-sandu-tao committed:
Update src/Parameters/ParameterTranslator.php

Co-Authored-By: Bartłomiej Marszał <bartlomiej@taotesting.com>

mike-ionut-mihai-sandu-tao committed:
Update src/Parameters/ParameterTranslator.php

Co-Authored-By: Bartłomiej Marszał <bartlomiej@taotesting.com>

mike-ionut-mihai-sandu-tao committed:
Update src/Parameters/ParameterTranslator.php

Co-Authored-By: Bartłomiej Marszał <bartlomiej@taotesting.com>

mike-ionut-mihai-sandu-tao committed:
Update src/SpannerStatement.php

Co-Authored-By: Bartłomiej Marszał <bartlomiej@taotesting.com>

mike-ionut-mihai-sandu-tao committed:
Update src/SpannerDriver.php

Co-Authored-By: Bartłomiej Marszał <bartlomiej@taotesting.com>

bartlomiejmarszal reviewed (approved):
My comments are postponed due to a large amount of code in this PR. We will apply my comments regarding code style in additional PR to split this commit into more readable smaller parts.
